php/class-customize-concurrency.php

```diff
@@ -14,4 +14,204 @@
+	 * @return array
+	 */
+	public function preserve_inserted_post_name( $data, $postarr ) {
+		if ( self::POST_TYPE === $data['post_type'] && isset( $postarr['post_name'] ) ) {
+			$data['post_name'] = wp_slash( $postarr['post_name'] );
+		}
+		return $data;
+	}
+
+	public function customize_save ( \WP_Customize_Manager $wp_customize ) {
+
+		$customizer_session_timestamp = isset( $_POST['customizer_session_timestamp'] ) ? intval( $_POST['customizer_session_timestamp'] ) : time();
+
+		$post_values = $wp_customize->unsanitized_post_values();
+		$setting_ids = array_keys( $post_values );
+		$invalidities = array();
+		$customize_saved_setting_posts = new \WP_Query( array(
+			'post_name__in' => $setting_ids,
+			'update_post_meta_cache' => false,
+			'update_post_term_cache' => false,
+			'post_type' => self::POST_TYPE,
+			'posts_per_page' => -1,
+			'ignore_sticky_posts' => true,
+		) );
+
+		foreach ( $customize_saved_setting_posts->posts as $setting_post ) {
+			$setting_id = $setting_post->post_name;
+			$existing_value = json_decode( $setting_post->post_content, true );
+			$is_conflicted = (
+				strtotime( $setting_post->post_modified_gmt ) > $customizer_session_timestamp
+				&&
+				isset( $post_values[ $setting_id ] )
+				&&
+				$existing_value !== $post_values[ $setting_id ]
+			);
+			if ( $is_conflicted ) {
+				$invalidities[ $setting_id ] = new \WP_Error( 'concurrency_conflict', __( 'Concurrency conflict' ), array( 'their_value' => $existing_value ) );
+			}
+		}
+
+		if ( ! empty( $invalidities ) ) {
+			$exported_setting_validities = array_map( array( $wp_customize, 'prepare_setting_validity_for_js' ), $invalidities );
+			$invalid_setting_count = count( $exported_setting_validities );
+			$response = array(
+				'setting_validities' => $exported_setting_validities,
+				'message' => sprintf( _n( 'There is %s invalid setting.', 'There are %s invalid settings.', $invalid_setting_count ), number_format_i18n( $invalid_setting_count ) ),
+			);
+
+			/** This filter is documented in wp-includes/class-wp-customize-manager.php */
+			$response = apply_filters( 'customize_save_response', $response, $this );
+			wp_send_json_error( $response );
+		}
+
+	}
+
+	public function customize_save_after() {
+
+		$post_values = $this->customize_manager->unsanitized_post_values();
+		$setting_ids = array_keys( $post_values );
+
+		foreach ( $setting_ids as $setting_id ) {
+			$setting = $this->customize_manager->get_setting( $setting_id );
+			if ( ! $setting ) {
+				continue;
+			}
+
+			$post_data = array(
+				'post_type' => self::POST_TYPE,
+				'post_status' => 'draft',
+				'post_title' => $setting_id,
+				'post_name' => $setting_id,
+				'post_content_filtered' => wp_json_encode( $post_values[ $setting_id ] ),
+				'post_date' => current_time( 'mysql', 0 ),
+				'post_date_gmt' => current_time( 'mysql', 1 ),
+				'post_author' => get_current_user_id(),
+			);
+
+			$this->suspend_kses();
+				wp_insert_post( $post_data, true );
```

Comment:
Also, this will need to use `wp_update_post()` if there is already a post that has this `post_name`. Refer to https://github.com/xwp/wp-customize-snapshots/blob/68e8f34d5bcb8c73307eaa2e91e19824396251a5/php/class-post-type.php#L497-L557